File: ADWs/pixel_office_client.py
"""Tiny client for ADWs code to publish pixel-office events without blocking.

This runs in-process inside the scheduler / runner. It fires HTTP POSTs with a
short timeout and swallows any exception — visualisation is a best-effort
side-channel, never a failure mode.
"""
from __future__ import annotations

import json
import os
import threading
import urllib.error
from urllib.request import Request, urlopen


def _post_sync(url: str, payload: dict, token: str | None, timeout: float) -> None:
    data = json.dumps(payload).encode("utf-8")
    headers = {"Content-Type": "application/json"}
    if token:
        headers["X-Hook-Token"] = token
    req = Request(url, data=data, headers=headers, method="POST")
    try:
        with urlopen(req, timeout=timeout) as _:
            pass
    except (urllib.error.URLError, OSError, TimeoutError):
        pass


def post_event(payload: dict, timeout: float = 1.5) -> None:
    url = os.environ.get("EVONEXUS_DASHBOARD_URL", "http://127.0.0.1:8080")
    url = url.rstrip("/") + "/api/pixel-office/hook"
    token = os.environ.get("PIXEL_OFFICE_HOOK_TOKEN") or None
    t = threading.Thread(target=_post_sync, args=(url, payload, token, timeout), daemon=True)
    t.start()

File: dashboard/terminal-server/src/pixel-office-client.js
/* Lightweight, non-blocking pixel-office event emitter for the terminal-server. */
const http = require('node:http');
const https = require('node:https');

function postEvent(payload) {
  try {
    const baseUrl = process.env.EVONEXUS_DASHBOARD_URL || 'http://127.0.0.1:8080';
    const url = new URL('/api/pixel-office/hook', baseUrl);
    const body = Buffer.from(JSON.stringify(payload), 'utf-8');
    const lib = url.protocol === 'https:' ? https : http;
    const opts = {
      hostname: url.hostname,
      port: url.port || (url.protocol === 'https:' ? 443 : 80),
      path: url.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': body.length,
        ...(process.env.PIXEL_OFFICE_HOOK_TOKEN ? { 'X-Hook-Token': process.env.PIXEL_OFFICE_HOOK_TOKEN } : {}),
      },
      timeout: 1500,
    };
    const req = lib.request(opts);
    req.on('error', () => {});
    req.on('timeout', () => req.destroy());
    req.write(body);
    req.end();
  } catch (_) { /* swallow — visualization is best-effort */ }
}

module.exports = { postEvent };
